### PR TITLE
Use GtkEventController for input method

### PR DESCRIPTION
**Pull request by ghost**
This is continued from https://github.com/wxWidgets/wxWidgets/pull/1357

When we type CJK, the input method uses preedit.
So when we click the mouse button, input method should be reset if needed.
When the input method is reset, the input method determines whether to commit the preedit or not. 
Using `gtk_entry_im_context_filter_keypress()` directly does not change `priv->need_im_reset` inside `GtkEntry`.
After many trial and error, I found the solution!
It is use to `gtk_event_controller_handle_event()` instead of `gtk_entry_im_context_filter_keypress()`.
I believe this is the final solution.
Now when we click the mouse button, GtkEntry resets im context as expected.

### CODE REVIEW

File: include/wx/gtk/textctrl.h [hunk 1]
```diff
@@ -10,6 +10,11 @@
 #ifndef _WX_GTK_TEXTCTRL_H_
 #define _WX_GTK_TEXTCTRL_H_
 
+#include <gtk/gtk.h>
```
Comment (vadz) [inline include/wx/gtk/textctrl.h]: We shouldn't include GTK+ headers from our own public headers. This means, of course, that we can't use `GTK_CHECK_VERSION` in them neither, but this is not really a problem, we can just use `__WXGTK3__` as the guard here and forward declare `GtkEventController`.

File: include/wx/gtk/textctrl.h [hunk 2]
```diff
@@ -10,6 +10,11 @@
...
+#if GTK_CHECK_VERSION(3, 24, 0)
+#include <gtk/gtkunixprint.h>
```
Comment (vadz) [inline include/wx/gtk/textctrl.h]: This header seems wrong, why should we need printing stuff here?

**Review by vadz — COMMENTED** [review]
Thanks, I didn't know about `GtkEventController`, but if it does what we need while still preserving the existing functionality, I definitely have no objections to using it. @paulcor Do you know more about this and do you have any comments about using this class?

We need to address the comments below/fix the build failures before merging this however, could you please do this? TIA!

File: include/wx/gtk/textctrl.h (hunk 2 above)
Comment (cogniti) [reply include/wx/gtk/textctrl.h]: You're right.
Only `#include <gtk/gtk.h>` is required.
https://developer.gnome.org/gtk3/stable/GtkEventControllerKey.html
https://developer.gnome.org/gtk3/3.24/GtkEventControllerKey.html

My devhelp manual is wrong.
Screenshot from 2019-06-25 10-12-19

**Comment by paulcor** [comment]
AFAICT there is no IM set on the controller, and no signal handlers connected to it, so I don't see how it could be doing much of anything. Am I missing something? Additionally, the controller appears to be expecting to also get key release events, so passing only press events looks incorrect.

**Comment by cogniti** [comment]
See:
https://gitlab.gnome.org/GNOME/gtk/blob/master/gtk/gtkeventcontrollerkey.c

```
static gboolean
gtk_event_controller_key_handle_event (GtkEventController *controller,
                                       const GdkEvent     *event)
{
...

  if (event_type != GDK_KEY_PRESS && event_type != GDK_KEY_RELEASE)
    return FALSE;

  if (key->im_context &&
      gtk_im_context_filter_keypress (key->im_context, (GdkEventKey *) event))
    {
      g_signal_emit (controller, signals[IM_UPDATE], 0);
      return TRUE;
...
}
```

```
static void
gtk_event_controller_key_class_init (GtkEventControllerKeyClass *klass)
{
...
  controller_class->handle_event = gtk_event_controller_key_handle_event;
...
}
```

I am afraid that the API document is not complete.
There is also `gtk_event_controller_key_get_im_context()` in the `gtkeventcontrollerkey.c`.
My mind was urgent. I did not consider `GtkPropagationPhase`.
https://developer.gnome.org/gtk3/stable/GtkEventController.html#GtkPropagationPhase.
Let's wait until `GtkEventControllerKey` API documentation is complete. After that, I will write code again considering `GtkPropagationPhase`.
Thank you for your time.

**Comment by paulcor** [comment]
> See:
https://gitlab.gnome.org/GNOME/gtk/blob/master/gtk/gtkeventcontrollerkey.c

Umm, yeah. Looking at the source code is how I arrived at the comment I made above.

> I did not consider `GtkPropagationPhase`

Probably not relevant. wxGTK doesn't use `GtkEventController`, and won't unless/until support for GTK4 is added. So plugging one in somewhere isn't likely to fix anything.

> Let's wait until `GtkEventControllerKey` API documentation is complete.

Probably not going to happen for GTK3.

**Comment by cogniti** [comment]
wxWidgets' `gtk_window_key_press_callback()` is connected to `"key-press-event"`.
`event_type` is `GDK_KEY_PRESS`. It calls `gtk_entry_im_context_filter_keypress()` or `gtk_text_view_im_context_filter_keypress()`.

Therefore I think there is no problem sending only `"key-press-event"` to `gtk_event_controller_key_handle_event()`.
There is also `gtk_im_context_filter_keypress()` inside the function.

I respect your opinion.
Thanks for your time.

```
static gboolean
gtk_event_controller_key_handle_event (GtkEventController *controller,
                                       const GdkEvent     *event)
{
  GtkEventControllerKey *key = GTK_EVENT_CONTROLLER_KEY (controller);
  GdkEventType event_type = gdk_event_get_event_type (event);
  GdkModifierType state;
  guint16 keycode;
  guint keyval;
  gboolean handled = FALSE;

  if (event_type == GDK_FOCUS_CHANGE)
    {
      gboolean focus_in;
      GdkCrossingMode mode;
      GdkNotifyType detail;

      gdk_event_get_focus_in (event, &focus_in);
      gdk_event_get_crossing_mode (event, &mode);
      gdk_event_get_crossing_detail (event, &detail);

      update_focus (key, focus_in, detail);

      key->current_event = event;

      if (focus_in)
        g_signal_emit (controller, signals[FOCUS_IN], 0, mode, detail);
      else
        g_signal_emit (controller, signals[FOCUS_OUT], 0, mode, detail);

      key->current_event = NULL;

      return FALSE;
    }

  if (event_type != GDK_KEY_PRESS && event_type != GDK_KEY_RELEASE)
    return FALSE;

  if (key->im_context &&
      gtk_im_context_filter_keypress (key->im_context, (GdkEventKey *) event))
    {
      g_signal_emit (controller, signals[IM_UPDATE], 0);
      return TRUE;
    }

  key->current_event = event;

  gdk_event_get_state (event, &state);
  if (key->state != state)
    {
      gboolean unused;

      key->state = state;
      g_signal_emit (controller, signals[MODIFIERS], 0, state, &unused);
    }

  gdk_event_get_keycode (event, &keycode);
  gdk_event_get_keyval (event, &keyval);

  if (event_type == GDK_KEY_PRESS)
    {
      g_signal_emit (controller, signals[KEY_PRESSED], 0,
                     keyval, keycode, state, &handled);
      if (handled)
        g_hash_table_add (key->pressed_keys, GUINT_TO_POINTER (keyval));
    }
  else if (event_type == GDK_KEY_RELEASE)
    {
      g_signal_emit (controller, signals[KEY_RELEASED], 0,
                     keyval, keycode, state);

      handled = g_hash_table_lookup (key->pressed_keys, GUINT_TO_POINTER (keyval)) != NULL;
      g_hash_table_remove (key->pressed_keys, GUINT_TO_POINTER (keyval));
    }
  else
    handled = FALSE;

  key->current_event = NULL;

  return handled;
}
```

**Pull request closed without merging**